### ISSUE TITLE
Embedded tomcat container configured with HTTP&HTTPS fails to start when spring-boot-starter-actuator is used

### ISSUE THREAD

**Issue by Laurenzi** [spring-projects/spring-boot#6193]
In Spring Boot 1.3.4 (also in 1.3.5) and using spring-boot-starter-actuator with embedded Tomcat (having JavaConfig for HTTP connector and autoconfigured HTTPS via application.properties) Spring Boot is unable to start embedded tomcat container.

Error message:

> org.springframework.context.ApplicationContextException: Unable to start embedded container; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'embeddedServletContainerFactory': Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.example.HttpsConfig$1]: No default constructor found; nested exception is java.lang.NoSuchMethodException: com.example.HttpsConfig$1.<init>()

Here is application.properties:

> management.port=9001
server.port.http=8080
server.port=8443

And here is the custom config for the other connector. Please notice for brevity, only HTTP connector is configured in the JavaConfig.

> 

	@Configuration
	public class HttpsConfig {

		@Value("${server.port}")
		private int httpsPort;

		@Value("${server.port.http}")
		private int httpPort;

		@Bean
		public EmbeddedServletContainerFactory servletContainer() {
			TomcatEmbeddedServletContainerFactory tomcat = new TomcatEmbeddedServletContainerFactory() {
			};
			tomcat.addAdditionalTomcatConnectors(initiateHttpConnector());
			return tomcat;
		}

		// redirect from httpPort -> httpsPort
		public Connector initiateHttpConnector() {
			Connector connector = new Connector("org.apache.coyote.http11.Http11NioProtocol");
			connector.setScheme("http");
			connector.setPort(httpPort);
			connector.setSecure(false);
			connector.setRedirectPort(httpsPort); // redirect to httpsPort
			return connector;
		}
	}

Removing `management.port` property from `application.properties` solves the issue. Problem is not present in Spring Boot 1.3.3 or earlier.

See [example project](https://github.com/Laurenzi/spring-boot-issues/tree/master/gh-6193) to reproduce the error.

**Comment by wilkinsona** [comment]
The change in behaviour is caused by https://github.com/spring-projects/spring-boot/commit/a1284bce61e6391b8c6dcb4e5e4d97612a32ded1 which fixes #5474.

Your `servletContainer` bean method is creating an anonymous `TomcatEmbeddedServletContainerFactory` subclass. Anonymous subclasses cannot be instantiated which prevents the embedded container for the management port from being created. The code you've shared doesn't need to use a subclass. Changing it as follows fixes the problem:

```diff
diff --git a/gh-6193/src/main/java/com/example/HttpsConfig.java b/gh-6193/src/main/java/com/example/HttpsConfig.java
index 94c0b78..21299ee 100644
--- a/gh-6193/src/main/java/com/example/HttpsConfig.java
+++ b/gh-6193/src/main/java/com/example/HttpsConfig.java
@@ -18,8 +18,7 @@ public class HttpsConfig {

     @Bean
     public EmbeddedServletContainerFactory servletContainer() {
-        TomcatEmbeddedServletContainerFactory tomcat = new TomcatEmbeddedServletContainerFactory() {
-        };
+        TomcatEmbeddedServletContainerFactory tomcat = new TomcatEmbeddedServletContainerFactory();
         tomcat.addAdditionalTomcatConnectors(initiateHttpConnector());
         return tomcat;
     }
```

Alternatively, if you've omitted some methods on the subclass for brevity, you can still fix the problem using an inner-class as long as it's named and static.

While the change in behaviour is unfortunate, I think it's worth it as it's straightforward to avoid it and it ensures that the correct container is used in the event of multiple embedded containers being on the classpath.

**Comment by wilkinsona** [comment]
Perhaps we can improve the diagnostics
